File: src/Cocoar.SignalARRR.Swift/Sources/CocoarSignalARRR/JsonHubProtocol.swift
import Foundation

/// ASCII Record Separator — SignalR message delimiter.
let recordSeparator: UInt8 = 0x1e

// MARK: - Hub Message Types

enum HubMessageType: Int {
    case invocation = 1
    case streamItem = 2
    case completion = 3
    case streamInvocation = 4
    case cancelInvocation = 5
    case ping = 6
    case close = 7
}

/// A parsed SignalR hub protocol message.
enum HubMessage {
    case invocation(target: String, arguments: [Any], invocationId: String?)
    case streamItem(invocationId: String, rawData: Data)
    case completion(invocationId: String, error: String?, rawData: Data?)
    case ping
    case close(error: String?)
}

// MARK: - JSON Hub Protocol

struct JsonHubProtocol {

    // MARK: Writing

    func writeHandshakeRequest() -> Data {
        Data(#"{"protocol":"json","version":1}"#.utf8) + [recordSeparator]
    }

    func writeInvocation(target: String, arguments: [Any], invocationId: String? = nil) throws -> Data {
        var msg: [String: Any] = [
            "type": HubMessageType.invocation.rawValue,
            "target": target,
            "arguments": try arguments.map { try serializeValue($0) }
        ]
        if let id = invocationId { msg["invocationId"] = id }
        return try toWire(msg)
    }

    func writeStreamInvocation(target: String, arguments: [Any], invocationId: String) throws -> Data {
        let msg: [String: Any] = [
            "type": HubMessageType.streamInvocation.rawValue,
            "invocationId": invocationId,
            "target": target,
            "arguments": try arguments.map { try serializeValue($0) }
        ]
        return try toWire(msg)
    }

    func writeCompletion(invocationId: String, result: Any? = nil, error: String? = nil) throws -> Data {
        var msg: [String: Any] = [
            "type": HubMessageType.completion.rawValue,
            "invocationId": invocationId
        ]
        if let error = error {
            msg["error"] = error
        } else if let result = result {
            msg["result"] = try serializeValue(result)
        } else {
            msg["result"] = NSNull()
        }
        return try toWire(msg)
    }

    func writeCancelInvocation(invocationId: String) throws -> Data {
        try toWire([
            "type": HubMessageType.cancelInvocation.rawValue,
            "invocationId": invocationId
        ])
    }

    func writePing() -> Data {
        Data(#"{"type":6}"#.utf8) + [recordSeparator]
    }

    // MARK: Parsing

    /// Parse the handshake response. Returns (error, remainingData).
    func parseHandshake(_ data: Data) throws -> (error: String?, remaining: Data?) {
        guard let sepIndex = data.firstIndex(of: recordSeparator) else {
            throw SignalRError.handshakeFailed("Incomplete handshake response")
        }
        let chunk = data[data.startIndex..<sepIndex]
        let json: [String: Any]
        if chunk.isEmpty {
            json = [:]
        } else {
            json = (try? JSONSerialization.jsonObject(with: Data(chunk))) as? [String: Any] ?? [:]
        }
        let error = json["error"] as? String

        let remainingStart = data.index(after: sepIndex)
        let remaining: Data? = remainingStart < data.endIndex ? Data(data[remainingStart...]) : nil
        return (error, remaining)
    }

    /// Parse one or more messages from wire data (split by record separator).
    func parseMessages(_ data: Data) throws -> [HubMessage] {
        data.split(separator: recordSeparator).compactMap { chunk in
            guard !chunk.isEmpty else { return nil }
            return try? parseMessage(Data(chunk))
        }
    }

    // MARK: Private

    private func parseMessage(_ data: Data) throws -> HubMessage? {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let typeRaw = dict["type"] as? Int,
              let type = HubMessageType(rawValue: typeRaw) else {
            return nil
        }
        switch type {
        case .invocation:
            return .invocation(
                target: dict["target"] as? String ?? "",
                arguments: dict["arguments"] as? [Any] ?? [],
                invocationId: dict["invocationId"] as? String
            )
        case .streamItem:
            return .streamItem(
                invocationId: dict["invocationId"] as? String ?? "",
                rawData: data
            )
        case .completion:
            return .completion(
                invocationId: dict["invocationId"] as? String ?? "",
                error: dict["error"] as? String,
                rawData: dict.keys.contains("result") ? data : nil
            )
        case .ping:
            return .ping
        case .close:
            return .close(error: dict["error"] as? String)
        case .streamInvocation, .cancelInvocation:
            return nil
        }
    }

    private func toWire(_ obj: [String: Any]) throws -> Data {
        try JSONSerialization.data(withJSONObject: obj) + [recordSeparator]
    }

    // MARK: Value Serialization

    func serializeValue(_ value: Any) throws -> Any {
        if value is NSNull { return NSNull() }

        // Unwrap Optional wrapped in Any
        let mirror = Mirror(reflecting: value)
        if mirror.displayStyle == .optional {
            guard let child = mirror.children.first else { return NSNull() }
            return try serializeValue(child.value)
        }

        // Primitives — JSONSerialization handles these directly
        if value is String || value is Bool { return value }
        if value is Int || value is Int8 || value is Int16 || value is Int32 || value is Int64 { return value }
        if value is UInt || value is UInt8 || value is UInt16 || value is UInt32 || value is UInt64 { return value }
        if value is Double || value is Float { return value }
        if value is NSNumber { return value }

        // Collections
        if let array = value as? [Any] { return try array.map { try serializeValue($0) } }
        if let dict = value as? [String: Any] { return try dict.mapValues { try serializeValue($0) } }

        // Encodable — encode to JSON then parse back to foundation object
        if let encodable = value as? any Encodable {
            let data = try JSONEncoder().encode(AnyEncodableBox(encodable))
            return try JSONSerialization.jsonObject(with: data, options: .fragmentsAllowed)
        }

        throw SignalRError.serializationFailed("Cannot serialize \(type(of: value))")
    }
}

// MARK: - Type-erased Encodable wrapper

struct AnyEncodableBox: Encodable {
    private let _encode: (Encoder) throws -> Void
    init(_ value: any Encodable) { _encode = value.encode }
    func encode(to encoder: Encoder) throws { try _encode(encoder) }
}

File: src/Cocoar.SignalARRR.Kotlin/signalarrr/src/main/kotlin/dev/cocoar/signalarrr/protocol/JsonHubProtocol.kt
package dev.cocoar.signalarrr.protocol

import dev.cocoar.signalarrr.HubProtocolKind
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject

/** The SignalR JSON hub protocol: one JSON object per message, terminated by the record separator. */
public class JsonHubProtocol : HubProtocol {
    override val kind: HubProtocolKind = HubProtocolKind.JSON
    override val isBinary: Boolean = false

    private val json = Json

    override fun writeHandshakeRequest(): ByteArray =
        toWire("{\"protocol\":\"json\",\"version\":1}")

    override fun parseHandshake(data: ByteArray): HandshakeResult = parseTextHandshake(data)

    override fun writeInvocation(target: String, arguments: List<JsonElement>, invocationId: String?): ByteArray {
        val fields = linkedMapOf<String, JsonElement>(
            "type" to JsonPrimitive(HubMessageType.INVOCATION.code),
        )
        if (invocationId != null) fields["invocationId"] = JsonPrimitive(invocationId)
        fields["target"] = JsonPrimitive(target)
        fields["arguments"] = JsonArray(arguments)
        return toWire(JsonObject(fields).toString())
    }

    override fun writeStreamInvocation(target: String, arguments: List<JsonElement>, invocationId: String): ByteArray {
        val obj = JsonObject(
            linkedMapOf(
                "type" to JsonPrimitive(HubMessageType.STREAM_INVOCATION.code),
                "invocationId" to JsonPrimitive(invocationId),
                "target" to JsonPrimitive(target),
                "arguments" to JsonArray(arguments),
            ),
        )
        return toWire(obj.toString())
    }

    override fun writeCompletion(invocationId: String, result: JsonElement?, error: String?): ByteArray {
        val fields = linkedMapOf<String, JsonElement>(
            "type" to JsonPrimitive(HubMessageType.COMPLETION.code),
            "invocationId" to JsonPrimitive(invocationId),
        )
        if (error != null) {
            fields["error"] = JsonPrimitive(error)
        } else {
            fields["result"] = result ?: JsonNull
        }
        return toWire(JsonObject(fields).toString())
    }

    override fun writeCancelInvocation(invocationId: String): ByteArray {
        val obj = JsonObject(
            linkedMapOf(
                "type" to JsonPrimitive(HubMessageType.CANCEL_INVOCATION.code),
                "invocationId" to JsonPrimitive(invocationId),
            ),
        )
        return toWire(obj.toString())
    }

    override fun writePing(): ByteArray = toWire("{\"type\":6}")

    override fun parseMessages(data: ByteArray): List<HubMessage> {
        val messages = mutableListOf<HubMessage>()
        var start = 0
        for (i in data.indices) {
            if (data[i] == HubProtocol.RECORD_SEPARATOR) {
                if (i > start) parseMessage(String(data, start, i - start, Charsets.UTF_8))?.let(messages::add)
                start = i + 1
            }
        }
        // A trailing chunk without a separator is not a complete message; SignalR always terminates.
        return messages
    }

    private fun parseMessage(text: String): HubMessage? {
        val obj = runCatching { json.parseToJsonElement(text).jsonObject }.getOrNull() ?: return null
        val typeCode = (obj["type"] as? JsonPrimitive)?.intOrNull ?: return null
        return when (HubMessageType.fromCode(typeCode)) {
            HubMessageType.INVOCATION -> HubMessage.Invocation(
                target = obj.string("target") ?: "",
                arguments = (obj["arguments"] as? JsonArray)?.toList() ?: emptyList(),
                invocationId = obj.string("invocationId"),
            )
            HubMessageType.STREAM_ITEM -> HubMessage.StreamItem(
                invocationId = obj.string("invocationId") ?: "",
                item = obj["item"] ?: JsonNull,
            )
            HubMessageType.COMPLETION -> HubMessage.Completion(
                invocationId = obj.string("invocationId") ?: "",
                error = obj.string("error"),
                result = obj["result"]?.takeUnless { it is JsonNull },
            )
            HubMessageType.PING -> HubMessage.Ping
            HubMessageType.CLOSE -> HubMessage.Close(
                error = obj.string("error"),
                allowReconnect = (obj["allowReconnect"] as? JsonPrimitive)?.booleanOrNull ?: false,
            )
            HubMessageType.STREAM_INVOCATION, HubMessageType.CANCEL_INVOCATION, null -> null
        }
    }

    private fun JsonObject.string(key: String): String? =
        (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

    private fun toWire(text: String): ByteArray = text.toByteArray(Charsets.UTF_8) + HubProtocol.RECORD_SEPARATOR
}
